### PR TITLE
#24609 continued johanandren

### LINKED ISSUE #24587: PatternCS.ask overload taking a function from ActorRef not selected by the compiler
``` java
import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.pattern.PatternsCS;
import akka.util.Timeout;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public class JavaMain {

    public static class ActorA extends AbstractActor {

        public ActorA(final ActorRef actorB) {
            CompletionStage<Object> response = PatternsCS.ask(
                    actorB,
                    replyTo -> new ActorB.Request("hello", replyTo),
                    Timeout.apply(1, TimeUnit.SECONDS));
            response.whenComplete((o, e) -> {
                if (o != null) System.out.println(o);
                if (e != null) System.out.println(e);
            });
        }

        @Override
        public Receive createReceive() {
            return emptyBehavior();
        }
    }

    public static class ActorB extends AbstractActor {

        public static class Request {

            public Request(final String text, ActorRef replyTo) {
                this.text = text;
                this.replyTo = replyTo;
            }

            public final String text;
            public final ActorRef replyTo;
        }

        public Receive createReceive() {
            return receiveBuilder()
                    .match(Request.class, request -> request.replyTo.tell(request.text, getSelf()))
                    .build();
        }
    }

    public static void main(String[] args) {
        final ActorSystem system = ActorSystem.create();
        final ActorRef actorB = system.actorOf(Props.create(ActorB.class), "actorB");
        final ActorRef actorA = system.actorOf(Props.create(ActorA.class, actorB), "actorA");
    }
}
```

Compiling with Java 8 and Java 9 and running results in:

```
akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://default/user/actorB#1415804947]] after [1000 ms]. Sender[null] sent message of type "akka.pattern.PatternsCS$$$Lambda$482/1291329720".
```

So obviousliy instead of selecting the "special" ask overload the "normal" one is selected and the function is not applied but instead sent as message.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #24587

### CODE REVIEW

**Comment by johanandren** [comment]
Also noticed the samples are old/incorrect and that the PatternCS stuff doesn't have tests (even if they just delegate, which is how we could miss this signature not being accessible) so created #24881 to track that.

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala [hunk 1]
```diff
@@ -388,18 +392,18 @@ object PatternsCS {
...
    * {{{
-   * final CompletionStage<Object> f = Patterns.ask(
+   * final CompletionStage<Object> f = PatternsCS.ask(
    *   worker,
-   *   new akka.japi.Function<ActorRef, Object> {
-   *     Object apply(ActorRef askSender) {
-   *       return new Request(askSender);
-   *     }
-   *   },
+   *   (replyTo) -> new Request(askSender),
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/Patterns.scala]: shouldn't `askSender` be `replyTo`

and no need for the `( )`

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala [hunk 2]
```diff
@@ -465,24 +469,6 @@ object PatternsCS {
...
-   *   selection,
-   *   new akka.japi.Function<ActorRef, Object> {
-   *     Object apply(ActorRef askSender) {
-   *       return new Request(askSender);
-   *     }
-   *   },
-   *   timeout);
-   * }}}
-   */
-  def ask(selection: ActorSelection, messageFactory: japi.Function[ActorRef, Any], timeoutMillis: Long): CompletionStage[AnyRef] =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/Patterns.scala]: why was this removed and not replaced?
we have one for `Future`:
```
def ask(selection: ActorSelection, messageFactory: japi.Function[ActorRef, Any], timeoutMillis: Long): Future[AnyRef]
```

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala [hunk 3]
```diff
@@ -347,9 +347,13 @@ object PatternsCS {
...
    * }}}
+   *
+   * @param actor          the actor to be asked
+   * @param messageFactory function taking an actor ref and returning the message to be sent
+   * @param timeout        the timeout for the response before failing the returned completion stage
    */
-  def ask(actor: ActorRef, messageFactory: japi.Function[ActorRef, Any], timeout: Timeout): CompletionStage[AnyRef] =
-    scalaAsk(actor, messageFactory.apply _)(timeout).toJava.asInstanceOf[CompletionStage[AnyRef]]
+  def askWithReplyTo(actor: ActorRef, messageFactory: japi.function.Function[ActorRef, Any], timeout: Timeout): CompletionStage[AnyRef] =
+    akka.pattern.extended.ask(actor, messageFactory.apply(_))(timeout).toJava.asInstanceOf[CompletionStage[AnyRef]]
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/pattern/Patterns.scala]: Don't we have the same problem with the ones returning `Future`, i.e. in `Patterns` (still Java API)

**Review by patriknw — CHANGES_REQUESTED** [review]

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala (hunk 3 above)
Comment (johanandren) [reply akka-actor/src/main/scala/akka/pattern/Patterns.scala]: Yes, I think you are right, I'll investigate.

File: akka-actor/src/main/scala/akka/pattern/Patterns.scala (hunk 2 above)
Comment (johanandren) [reply akka-actor/src/main/scala/akka/pattern/Patterns.scala]: This signature version, `timeoutMillis` was added above, but Heiko forgot to remove this one.

Comment (johanandren) [reply akka-actor/src/main/scala/akka/pattern/Patterns.scala]: Oh wait, they are in different objects. I'll fix.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
I manually squashed this into 2 commits, let's keep it like that so that credit is given to both authors.

**Comment by patriknw** [comment]
I will merge this after validation so that it can be included in tomorrow's release, I think enough time for review has been given.

**Pull request merged**